archeryutils/handicaps/handicap_scheme.py

```diff
@@ -0,0 +1,764 @@
+"""Code for a handicap scheme class performing calculations for a  generic handicap scheme.
+
+Extended Summary
+----------------
+Calculates arrow and round scores for a variety of target faces of
+given distance and diameter:
+
+- 5-zone
+- 10-zone
+- 10-zone 6-ring
+- 10-zone compound
+- 10-zone 5-ring
+- 10-zone 5-ring compound
+- WA_field
+- IFAA_field
+- Beiter-hit-miss
+- Worcester
+- Worcester 2-ring
+
+References
+----------
+- The construction of the graduated handicap tables for target
+  archery
+  Lane, D (2013)
+  https://www.jackatkinson.net/files/Handicap_Tables_2013.pdf
+- New AGB: Atkinson, J
+- Modelling archers’ scores at different distances to quantify
+  score loss due to equipment selection and technique errors
+  Park, J (2014)
+  https://doi.org/10.1177%2F1754337114539308
+
+"""
+
+import warnings
+from abc import ABC, abstractmethod
+from typing import Optional, TypeVar, Union, overload
+
+import numpy as np
+import numpy.typing as npt
+
+from archeryutils import rounds, targets
+
+FloatArray = TypeVar("FloatArray", float, npt.NDArray[np.float_])
+
+
+class HandicapScheme(ABC):
+    """
+    Abstract Base Class to represent a generic handicap scheme.
+
+    Attributes
+    ----------
+    name : str
+        The name of the handicap scheme.
+    arw_d_out : float
+        diameter of an outdoor arrow [metres] for this scheme, default 5.5e-3
+    arw_d_in: float
+        diameter of an indoor arrow [metres] for this scheme, default 9.3e-3
+    desc_scale: bool
+        does the scheme work on a descending scale i.e. lower handicap is better, default True
+
+    Methods
+    -------
+    sigma_t(handicap, dist)
+        Calculate angular deviation for given handicap and distance.
+    sigma_r(handicap, dist)
+        Calculate radial deviation for a given handicap and distance.
+    arrow_score(target, handicap, arw_d=None)
+        Calculate the average arrow score for a given target and handicap.
+    score_for_passes(rnd, handicap, arw_d=None, rounded_score=True)
+        Calculate the expected score for all passes in a round for a given handicap.
+    score_for_round(rnd, handicap, arw_d=None, rounded_score=True)
+        Calculate the expected score for a round for a given handicap.
+    handicap_from_score(score,rnd, arw_d=None, int_prec=False)
+        Calculate the handicap for a given score on a given round.
```

Comment:
I think it might make more sense to settle on round as the first positional argument to `handicap_from_score`, given `score_for_passes` and `score_for_round` both take it first.

Reply:
Yes, one think I need to do as part of this MR is standardise variable names and order - I noticed as I consolidated that there were discrepancies between `round_score_up` and `rounded_score` etc. that both performed the same function.

Reply:
I'm inclined to make the standardised order `handicap`, `round`, as this would match the `sigma_r/t()` which uses `handicap`, `distance`. Not sure it's massively important as long as it gets standardised in some way.

Reply:
This _should_ be dealt with in https://github.com/jatkinson1000/archeryutils/pull/78/commits/851a98cc2d12233e1087da98216007033844d37e but please do cast an eye as I could easily have missed something.

Reply:
Actually, it would probably make sense to also standardise the classification functions to also take `score` then `round`? And I need to update the docs etc.

Reply:
Standardising classification equations to take score as first argument implemented in https://github.com/jatkinson1000/archeryutils/pull/78/commits/f30fc02d6ccf008e8f8b792695b25846b180d909

Are you happy with how things are now set up/standardised @TomHall2020?

Reply:
Consistent is good, played with it interactively and can settle/get used to it so all happy for me